File: rest-common/src/main/java/org/chronopolis/rest/constraints/E123.java
package org.chronopolis.rest.constraints;

import javax.validation.Constraint;
import javax.validation.Payload;
import java.lang.annotation.Documented;
import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.ANNOTATION_TYPE;
import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.ElementType.PARAMETER;
import static java.lang.annotation.ElementType.TYPE_USE;

/**
 * Annotation for Hibernate validation on Phone Numbers (with the help of libphonenumber)
 *
 * @author shake
 */
@Documented
@Target({FIELD, METHOD, PARAMETER, ANNOTATION_TYPE, TYPE_USE})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = CheckPhoneNumber.class)
@Repeatable(E123.List.class)
public @interface E123 {

    String message() default "Invalid PhoneNumber";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

    @Target({FIELD, METHOD, PARAMETER, ANNOTATION_TYPE, TYPE_USE})
    @Retention(RetentionPolicy.RUNTIME)
    @Documented
    @interface List {
        E123[] value();
    }
}

File: rest-models/src/main/kotlin/org/chronopolis/rest/constraints/E123.kt
package org.chronopolis.rest.constraints

import javax.validation.Constraint
import kotlin.reflect.KClass

/**
 * JSR-303 annotation for validation on phone numbers (with the help of libphonenumber)
 *
 * @author shake
 */
@MustBeDocumented
// @Target(AnnotationTarget.FIELD)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [PhoneNumberValidator::class])
annotation class E123(
        val message: String = "Invalid PhoneNumber",
        val groups: Array<KClass<out Any>> = [],
        val payload: Array<KClass<out Any>> = []
)